### PR TITLE
feat: support dubbo

### PR DESCRIPTION
**Pull request by spacewander**
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: doc/plugins/dubbo-proxy.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        "/hello"
+    ],
+    "plugins": {
+        "dubbo-proxy": {
+            "service_name": "org.apache.dubbo.sample.tengine.DemoService",
+            "service_version": "0.0.0",
+            "method": "tengineDubbo"
+        }
+    },
+    "upstream_id": 1
```
Comment (gxthrj) [inline doc/plugins/dubbo-proxy.md:91]: It seems to conflict with the design of the dubbo registry, which will confuse users.

File: t/lib/dubbo-backend/dubbo-backend-provider/src/main/resources/META-INF/spring/dubbo-demo-provider.xml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+       xmlns:dubbo="http://dubbo.apache.org/schema/dubbo"
+       xmlns="http://www.springframework.org/schema/beans"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.3.xsd
+       http://dubbo.apache.org/schema/dubbo http://dubbo.apache.org/schema/dubbo/dubbo.xsd">
+
+    <!-- provider's application name, used for tracing dependency relationship -->
+    <dubbo:application name="demo-provider"/>
+
+    <!-- use multicast registry center to export service -->
+    <dubbo:registry address="multicast://224.5.6.7:1234"/>
```
Comment (gxthrj) [inline t/lib/dubbo-backend/dubbo-backend-provider/src/main/resources/META-INF/spring/dubbo-demo-provider.xml:28]: Multicast is for test and develop environment, not for the production.

Comment (spacewander) [reply t/lib/dubbo-backend/dubbo-backend-provider/src/main/resources/META-INF/spring/dubbo-demo-provider.xml:28]: This part of code is just provided a test backend, it is not used for prod.

File: doc/plugins/dubbo-proxy.md (hunk 1 above)
Comment (spacewander) [reply doc/plugins/dubbo-proxy.md:91]: It seems neither Tengine or Envoy implement dubbo registry?

Comment (spacewander) [reply doc/plugins/dubbo-proxy.md:91]: We can implement it if user requires. This PR is big enough.

Comment (membphis) [reply doc/plugins/dubbo-proxy.md:91]: agree

File: doc/plugins/dubbo-proxy.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+{
+    "nodes": {
+        "127.0.0.1:20880": 1
+    },
+    "type": "roundrobin"
+}'
+
+curl http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
+{
+    "methods": ["GET"],
```
Comment (membphis) [inline doc/plugins/dubbo-proxy.md:80]: user example, less is better. so we can remove this line

**Review by membphis — APPROVED** [review]
LGTM

**Comment by membphis** [comment]
We need a way to avoid recompiling openresty every time in ci. We can implement this in a new PR.

Now it takes 20 minutes, which is a bit long.

image

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -59,6 +59,15 @@ jobs:
...
         run: sudo ./.travis/${{ matrix.os_name }}_runner.sh before_install
 
+      - name: Start Dubbo Backend
+        if: matrix.os_name == 'linux_openresty'
+        run: |
+          sudo apt install -y maven
+          cd t/lib/dubbo-backend
+          mvn package
+          cd dubbo-backend-provider/target
+          java -Djava.net.preferIPv4Stack=true -jar dubbo-demo-provider.one-jar.jar > /tmp/java.log &
```
Comment (nic-chen) [inline .github/workflows/build.yml:69]: LGTM

It is better to move `t/lib/dubbo-backend` to a separate repo, and only need to download the compiled jar when running CI.

Comment (spacewander) [reply .github/workflows/build.yml:69]: As the content of Java code is depended on what you want to test, it is inconvenient to move it to a separate repo.

Comment (membphis) [reply .github/workflows/build.yml:69]: `mvn package` + `java -Djava.net.preferIPv4Stack=true`

How long does it usually take?

Comment (spacewander) [reply .github/workflows/build.yml:69]: @membphis 
32 seconds

Comment (membphis) [reply .github/workflows/build.yml:69]: I think this is acceptable

File: README.md [hunk 5]
```diff
@@ -79,6 +79,7 @@ A/B testing, canary release, blue-green deployment, limit rate, defense against
 - **Multi protocols**
 
   - [TCP/UDP Proxy](doc/stream-proxy.md): Dynamic TCP/UDP proxy.
+  - [dubbo Proxy](doc/plugins/dubbo-proxy.md): Dynamic HTTP to dubbo proxy.
```
Comment (tokers) [inline README.md:82]: Maybe `Dubbo` is better?

File: t/plugin/dubbo-proxy/route.t [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                service_version: 0.0.0
+                method: hello
+        id: 1
+        upstream_id: 1
+--- response_body
+dubbo success
+
+
+
+=== TEST 5: work with comsumer
```
Comment (tokers) [inline t/plugin/dubbo-proxy/route.t:160]: Typo: consumer.

Comment (spacewander) [reply t/plugin/dubbo-proxy/route.t]: Fixed.

**Review by tokers — APPROVED** [review]

**Comment by wxbty** [comment]
@spacewander   Hi, can you please reply, will the dubbo backend service start automatically when running the test? I opened a PR #9660  that supports the dubbo protocol, and want to reuse the dubbo backend service, but by default, connection refused is returned.

**Pull request merged**